site/app/libraries/routers/WebRouter.php

```diff
@@ -62,11 +62,13 @@ public function __construct(Request $request, Core $core, $logged_in, $is_api =
                 // prevent user that is not logged in from going anywhere except AuthenticationController
                 if (!$this->logged_in &&
                     !Utils::endsWith($this->parameters['_controller'], 'AuthenticationController')) {
-                    $this->api_authorized = false;
+                    $this->core->getOutput()->renderJsonFail("Unauthorized access. Please log in.");
+                    die($this->core->getOutput()->getOutput());
                 }
             }
             catch (ResourceNotFoundException $e) {
-                $this->parameters = null;
+                $this->core->getOutput()->renderJsonFail("Endpoint not found.");
+                die($this->core->getOutput()->getOutput());
```

Comment:
Might not need to do it in this PR (probably better once we kill off the ClassicRouter for good), but we should try to make sure we're not inserting dies into code since it means we cannot really write a test now for this path which is not optimal. (I regret the die that exists in the Output class and look forward to the recent round of changes getting us closer to where we can probably remove it).

Reply:
Thanks for your suggestion. `die` is indeed bad for testing. After we have `Response` probably we can better handle that by having a bad response for both JSON and HTML (currently I move it here because I think it is for API only). I will remove `die`s next time. Let there be `die` in `redirect` only.

Reply:
For reference, I'd like to potentially think of adding a `RedirectResponse` class as well so that we can remove the die from the redirect function as well.